texlive.combine: add paperSize parameter for setting paper sizes

xworld21 opened the pull request:
## Description of changes
Add a new parameter `paperSize` to `texlive.combine` with for setting the default paper size, either for all programs at once (with a string), or with different values per program (with an attribute set).

We should probably come up with a test for this.

PS: I hope that calling a single function in the module `TeXLive::TLPaper` does not make Nixpkgs GPL2!

## Things done
- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [X] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

xworld21 committed:
texlive.combine: add paperSize parameter for setting paper sizes

veprbl commented on pkgs/tools/typesetting/tex/texlive/combine.nix:
```diff
@@ -5,10 +5,21 @@ args@{
+# - an attribute set like { dvipdfmx = "a4"; }, to set different page sizes
+#   for different programs
+# see tlmgr and $TEXMFROOT/tlpkg/TeXLive/TLPaper.pm for the supported programs
+# and page sizes
```
Should go to the docs?

xworld21 commented:
PPS: I am going to add several tests in a different PR, including one for the paper size.

apfelkuchen6 reviewed (commented):
Unfortunately this further complicates the upcoming format containers, but I can definitely see this being useful for some people (imho every document should specify the paper size explicitly, but different people have different workflows...) 

What are your ideas on how to proceed with the format containers?

xworld21 commented:
> Unfortunately this further complicates the upcoming format containers, but I can definitely see this being useful for some people (imho every document should specify the paper size explicitly, but different people have different workflows...)
> 
> What are your ideas on how to proceed with the format containers?

Good question, and the motivation for `paperSize` is in fact to make format containers more reliable. In TeX Live, a format must be rebuilt when:
- one of its core dependencies (`fmttriggers`) is updated: we don't need to think about this, because Nix enforces this by construction
- hyphenation patterns have changed (updated, added, removed); tlpdb specifies which language file is relevant (language.dat/def/dat.lua, or user config for context)
- the default paper size changes

I have a rough implementation of format containers that simply calls `texlive.combine` with a private argument `_buildFormatsFor`. If it contains a package, `texlive.combine` generates the formats of that package, then stops and deletes everything else (so you only get `/share/texmf-var/web2c//*.fmt`). If `_buildFormatsFor` is not specified, it calls itself recursively for each package containing formats, passing along `paperSize` and all language packages. (Bonus points if one filters `paperSize` so that changing the size for `pdftex` does not recompile formats for `dvipdfmx`.)

I say "rough" because this is cumbersome to do with `texlive.combine` so I am holding it off until we get `texlive.buildEnv`.

veprbl commented on doc/languages-frameworks/texlive.section.md:
```diff
@@ -28,6 +28,29 @@ Since release 15.09 there is a new TeX Live packaging that lives entirely under
+    paperSize = {
+      "dvipdfmx" = "legal";
+      "psutils" = "letter";
+    };
```
How do I set the default without trying to list every program in this case?

veprbl replied:
Is this even practical?

apfelkuchen6 replied:
Special-casing the `default`-attribute to also run `paper_all` is definitely possible. I can't really imagine a usecase for this, but it's probably more common than wanting to specify pairwise distinct paper sizes for all the programs (there can't be that many sizes, right?)

xworld21 replied:
Right. I was exposing the tlmgr API to the letter, but tlmgr can be run twice and `combine` cannot. I think `default` is very unambiguous and I can't think of anything better (`all` is ambiguous, `_default` is a bit ugly).

I agree that this is a fairly abstract concern – even Debian only offers a4/letter I think. Hence the `paperSize = "letter"` version.

(xdvi has tons of sizes for some reason! most of the others are just a4/letter.)

xworld21 replied:
Conclusion for today: `paper_all` returns an error when some of the packages are not installed, even though everything goes well. We should not ignore errors as they could represent actual failures, so we have to list the relevant packages by hand and configure only the ones that are installed.

I say, let's focus on `texlive.buildEnv` and come back to this later.

xworld21 commented:
Rearranging to be after #248960 because I am getting failures now!
